client/state/posts/revisions/reducer.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * External dependencies
+ */
+import { combineReducers } from 'redux';
+import { flow, keyBy, map } from 'lodash';
+import mapValues from 'lodash/fp/mapValues';
+import pick from 'lodash/fp/pick';
+
+/**
+ * Internal dependencies
+ */
+import {
+	POST_REVISIONS_RECEIVE,
+	POST_REVISIONS_REQUEST,
+	POST_REVISIONS_REQUEST_FAILURE,
+	POST_REVISIONS_REQUEST_SUCCESS,
+	SERIALIZE,
+	DESERIALIZE
+} from 'state/action-types';
+
+export function requesting( state = {}, action ) {
+	switch ( action.type ) {
+		case POST_REVISIONS_REQUEST:
+		case POST_REVISIONS_REQUEST_FAILURE:
+		case POST_REVISIONS_REQUEST_SUCCESS:
+			return {
+				...state,
+				[ action.siteId ]: {
+					...state[ action.siteId ],
+					[ action.postId ]: action.type === POST_REVISIONS_REQUEST,
+				},
+			};
+
+		case SERIALIZE:
+		case DESERIALIZE:
+			return {};
+	}
+
+	return state;
+}
+
+export function revisions( state = {}, action ) {
+	if ( action.type === POST_REVISIONS_RECEIVE ) {
+		const { siteId, postId } = action;
+		return {
+			...state,
+			[ siteId ]: {
+				...state[ siteId ],
+				[ postId ]: keyBy( map( action.revisions, normalizeRevisionForState ), 'id' )
+			}
+		};
+	}
+
+	return state;
+}
+
+export default combineReducers( {
+	requesting,
+	revisions,
+} );
+
+function normalizeRevisionForState( revision ) {
```

Comment:
I wonder if we should normalize the objects in the state or in the data middleware layer. I probably have a preference over the data layer (because we could imagine using different APIs with different shapes and reuse the same reducer)